pennylane/transforms/hamiltonian_expand.py

```diff
@@ -155,21 +161,29 @@ def hamiltonian_expand(tape, group=True):
 
             new_tape = new_tape.expand(stop_at=lambda obj: True)
             tapes.append(new_tape)
-    else:
-        coeffs = hamiltonian.coeffs
 
-        tapes = []
-        for o in hamiltonian.ops:
-            with tape.__class__() as new_tape:
-                for op in tape.operations:
-                    op.queue()
-                qml.expval(o)
+        def processing_fn(res_groupings):
+            dot_products = [
+                qml.math.dot(r_group, c_group)
+                for c_group, r_group in zip(coeff_groupings, res_groupings)
+            ]
+            return qml.math.sum(qml.math.stack(dot_products), axis=0)
 
-            tapes.append(new_tape)
+        return tapes, processing_fn
+
+    coeffs = hamiltonian.data
```

Comment:
Without grouping, the shapes are different:
`coeffs` looks like
`[<tf.Tensor: shape=(), dtype=float64, numpy=-0.05>, <tf.Tensor: shape=(), dtype=float64, numpy=0.17>, <tf.Tensor: shape=(), dtype=float64, numpy=1.2>]` and `res` like
`[<tf.Tensor: shape=(1,), dtype=float64, numpy=array([-0.12777055])>, <tf.Tensor: shape=(1,), dtype=float64, numpy=array([0.0166009])>, <tf.Tensor: shape=(1,), dtype=float64, numpy=array([1.])>]`

Reply:
PS: We could just add an extra dimension to the coeffs and use the same processing function, but I had such trouble figuring these shapes out, it is much coder friendly to treat the cases separate I think?